docs/development/adr/0018-bin-hist-reduction-dims.md

```diff
@@ -0,0 +1,123 @@
+# ADR 0018: More explicit control of replaced dims in `bin` and `hist` functions
+
+-   Status: proposed
+-   Deciders: Simon, Jan-Lukas, Sunyoung
+-   Date: 2024-07-12
+
+## Context
+
+### Background
+
+Currently functions such as `bin` and `hist` derive their behavior from looking at whether the coordinate being used for the operation is a "dimension-coordinate" of the data array, i.e., a coordinate with its name matching a dimension of the data.
+Recently we have seen more and more cases where this behavior leads to surprising and too implicit results.
+To control the actual behavior a number of workarounds in user-code are required, using combinations of functions such as `flatten`, `group`, `rename_dims`, `drop_coords`, `transpose`, etc.
+Furthermore, there are cases where the default behavior shows sub-optimal performance, which can be avoided using further workarounds.
+All of this is cumbersome and error prone, so this ADR proposes a more predictable and explicit control of the dimension(s) to be replaced or added.
+
+### Proposed solution
+
+1. Add a `dim` argument to all `bin` and `hist` functions to specify the dimension(s) to be replaced.
```

Comment:
I am not sure the name `dim` is explicit enough in what it is doing.
I don't know if we can pick another name like `remove` or something better?

I can see that in our reduction operations, using the argument `dim` often means that this dimension will disappear in the output, but we do have some other operations like `concat` where the dim will appear in the output.

Reply:
I think `dim` is best, since I feel it *is* consistent. `concat` is probably the only case where it is used for something else? But note that `sc.concat` and `da.bins.concat` both use `dim` for different things, for a function with the same name!

Reply:
I still think that `da.hist(z=50, dim='y')` is not so clear what it is doing.
In a reduction operation, I think it is clearer, as it is like the numpy `axis` argument: `da.min(dim='y')` is pretty clear that we apply the min along the `y` dim.

In `make_binned` we have `erase` as an argument. Would `erase` be clearer here?

That said, if we expect this to be used only by 'advanced' users, maybe the naming is not crucial?

Reply:
Have a look at NumPy, or the Array API standard. Arguments like this are always named `axis`. In Scipp we are (hopefully?) consistently naming them `dim`. I have always imagined that this would be least confusing, since no matter what function you use it is always `dim`, with no need to read the documentation (provided that you know or guess the function has such an argument).